index.js

```diff
@@ -89,6 +91,120 @@ async function create ({
   }
 }
 
+/**
+ * checkPeersSync.
+ *
+ * @description check if there is at least one peer fully synchronized (up to date) with our drive
+ * @async
+ * @param {Object} contentFeed hypercore content feed
+ * @param {Array} stats array containing { file: String, start: Number, end: Number}. start === stat.offset and end === blocks
+ * @param {[Number]} wait a number in miliseconds to use as a function timeout. Defaults to 120000 (2 minutes). OPTIONAL
+ */
+function checkPeersSync (contentFeed, stats, wait = 120000) {
+  if (!contentFeed) {
+    throw new Error('content feed is required')
+  }
+
+  const result = {}
+  result.promise = new Promise((resolve, reject) => {
+    const timer = setTimeout(() => {
+      result.reject(new Error(`timeout exceded: ${wait}ms`))
+    }, wait)
+
+    result.resolve = () => {
+      contentFeed.off('upload', checkSync)
+      clearTimeout(timer)
+      resolve()
+    }
+
+    result.reject = err => {
+      contentFeed.off('upload', checkSync)
+      clearTimeout(timer)
+      reject(err)
+    }
+  })
+
+  const checkSync = debounce(() => {
+    console.log({ peers: contentFeed.peers })
+    console.log({ stats })
+
+    for (const peer of contentFeed.peers) {
+      // Ignore peers without remote bitfields
+      if (!peer.remoteBitfield) continue
+
+      // iterate over ranges
+      let haveAll = true
+
+      for (const stat of stats) {
+        for (let idx = stat.start; idx <= stat.end; idx++) {
+          if (idx === 0) continue
+          if (!peer.remoteBitfield.get(idx)) {
+            haveAll = false
+            break
+          }
+        }
+
+        if (!haveAll) {
+          // check next peer
+          break
+        }
+      }
+
+      if (haveAll) {
+        // content synced with at least one peer
+        return result.resolve()
+      }
+    }
+  }, 150)
+
+  checkSync()
+  contentFeed.on('upload', checkSync)
+
+  return result.promise
+}
+
+/**
+ * updateStats.
+ *
+ * @description Iterates over each file of the drive tracking start and end of each item
+ * @async
+ * @param {Object} drive Hyperdrive instance
+ * @return {Array} An array containing [<{file:String, start:Number, end:Number}>]
+ */
+async function updateStats (drive) {
+  const list = await drive.readdir('/')
```

Comment:
Would it make sense to do a call to `drive.stats('/')` here instead? It does the recursive lookup for you already. https://github.com/hypercore-protocol/hyperdrive/blob/master/index.js#L890

Reply:
totally forget about `stats` method, yes, good idea mauve